mongosql/src/translator/expressions.rs

```diff
@@ -187,48 +187,55 @@ impl MqlTranslator {
+        // Set fields are fields that start with a '$' or contain a '.'.
+        let (normal_fields, set_fields): (Vec<_>, Vec<_>) = mir_document
+            .into_iter()
+            .partition(|(k, _)| !k.starts_with('$') && !k.contains('.'));
+
+        // normal fields are processed first since they do not require any special
+        // handling.
+        for (k, v) in normal_fields {
             if k.is_empty() {
                 return Err(Error::InvalidDocumentKey(k));
             }
 
-            // If a key starts with a '$' or contains a '.', we must use a
-            // SetField operator to represent this document. Every key-value
-            // pair must be inserted via SetFields, so once we encounter a
-            // key with these properties we start iteration over and build the
-            // SetField expression tree.
-            //
-            // This case is not likely to happen often, which is why we do not
-            // build the set_field_expr simultaneously. This set up allows us
-            // to iterate over the keys once and build the document expr without
-            // extra allocations for _most_ translation cases. If a key _does_
-            // contain a '$' or '.', this method could, in the worst case, take
-            // more time and/or memory, but that is a worthwhile tradeoff.
-            if k.starts_with('$') || k.contains('.') {
-                // Start with an empty document as the initial "input".
-                let mut set_field_expr = air::Expression::Document(UniqueLinkedHashMap::new());
-                for (k, v) in mir_document {
-                    let translated_v = self.translate_expression(v)?;
-                    set_field_expr = air::Expression::SetField(air::SetField {
-                        field: k.clone(),
-                        input: Box::new(set_field_expr),
-                        value: Box::new(translated_v),
-                    })
-                }
-                return Ok(set_field_expr);
-            }
-
             let translated_v = self.translate_expression(v)?;
-            doc_expr.insert(k, translated_v)?;
+            doc_expr.insert(k.to_string(), translated_v)?;
+        }
+
+        // If there are no set fields, we can return the document expression
+        if set_fields.is_empty() {
+            return Ok(air::Expression::Document(doc_expr));
         }
 
-        Ok(air::Expression::Document(doc_expr))
+        // set fields are required to be set inside a SetField expression.
+        // The document expression calculated above is the initial input to the
+        // SetField expression(s). The SetField expression is
+        // chained with other SetField expressions for each set field.
+        let mut set_field_expr = air::Expression::Document(doc_expr);
+        for (k, v) in set_fields {
+            if k.is_empty() {
+                return Err(Error::InvalidDocumentKey(k));
+            }
```

Comment:
I understand why we have this as a "just-in-case" but shouldn't this be impossible? The `partition` method separated any empty strings into the first group, `normal_fields`, leaving only strings that start with `$` or include `.` in this set.

Reply:
Updated to be handled during initial partitioning